server/plugin_release.go

```diff
@@ -230,170 +443,275 @@ func downloadAsset(assetURL, tempFolder string) (string, error) {
-		return err
+		return nil, fmt.Errorf("failed to setup client config err=%w", err)
 	}
 
-	LogInfo("will sign file from path " + filePath)
-	cmd := fmt.Sprintf("sudo -u signer /opt/plugin-signer/sign_plugin.sh %s", filePath)
-	stdout, stderr, err := sshClient.Run(cmd)
+	client, err := ssh.Dial("tcp", fmt.Sprintf("%v:22", cfg.PluginSigningSSHHost), &clientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup client config err=%w", err)
+	}
+
+	sftp, err := sftp.NewClient(client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup sftp client err=%w", err)
+	}
+
+	return sftp, nil
+}
+
+// hasAllPlatformBinaries verifies if plugin tar contains 3 platform binaries.
+func hasAllPlatformBinaries(filePath string) error {
+	f, err := os.Open(filePath)
 	if err != nil {
-		LogInfo(stdout)
-		LogInfo(stderr)
-		LogInfo(err.Error())
 		return err
 	}
-	cmd = fmt.Sprintf("rm -f %s", filePath)
-	stdout, stderr, err = sshClient.Run(cmd)
+	defer f.Close()
+
+	gzf, err := gzip.NewReader(f)
 	if err != nil {
-		LogInfo(stdout)
-		LogInfo(stderr)
-		LogInfo(err.Error())
 		return err
 	}
-	LogInfo(stdout)
-	LogInfo("Done signing the artifact")
+
+	tarReader := tar.NewReader(gzf)
+	serverDist := map[string]struct{}{}
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		name := header.Name
+		switch header.Typeflag {
+		case tar.TypeReg:
+			if strings.Contains(name, "plugin-linux-amd64") || strings.Contains(name, "plugin-windows-amd64.exe") || strings.Contains(name, "plugin-darwin-amd64") {
+				serverDist[name] = struct{}{}
+			}
+		}
+	}
+
+	if len(serverDist) != 3 {
+		return fmt.Errorf("plugin tar contains %+v, but should contain all platform binaries", serverDist)
+	}
+
 	return nil
 }
 
-func uploadSignedArtifcatsToS3(fileToUpload []string, tempFolder string) error {
-	LogInfo("Uploading signed assets to S3")
+// archiveContains returns filenames that matches a given string.
+func archiveContains(filePath string, contains string) ([]string, error) {
+	var result []string
+	f, err := os.Open(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open archive %s, err=%w", filePath, err)
+	}
+
+	gzf, err := gzip.NewReader(f)
+	if err != nil {
+		return nil, err
+	}
+
+	tarReader := tar.NewReader(gzf)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to read next %s, err=%w", filePath, err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeReg:
+			if strings.Contains(header.Name, contains) {
+				result = append(result, filepath.Base(header.Name))
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// privateKey Loads a private and public key from "path" and returns a SSH ClientConfig to authenticate with the server
+func privateKey(username string, path string, certPath string, keyCallBack ssh.HostKeyCallback) (ssh.ClientConfig, error) {
```

Comment:
I am kind of missing a verb in the function name, but naming is hard. Would `getSshClient` or `usePrivateKey` make sense?

Reply:
`createSSHCient`?